caproto/ioc_examples/dynamic_calc.py

```diff
@@ -0,0 +1,58 @@
+#!/usr/bin/env python3
+from caproto.server import pvproperty, PVGroup, ioc_arg_parser, run
+from caproto import ChannelDouble
+
+import ast
+
+
+def extract_names(code):
+    m = ast.parse(code)
+    c = compile(code, '', 'eval')
+    # TODO also skip np namespace
+    return c, set(n.id for n in ast.walk(m)
+                  if isinstance(n, ast.Name))
+
+
+class SmartCalc(PVGroup):
+    formula = pvproperty(value='')
+    proc = pvproperty(value=0)
+    output = pvproperty(value=0.0)
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._pvdb = {}
+        self.out_db = {}
+        self.out_db.update(self.pvdb)
+
+    @formula.putter
+    async def formula(self, instance, value):
+        c, names = extract_names(value)
+
+        self.c = c
+        self.names = names
+
+        self._pvdb.clear()
+        self._pvdb.update({f'{self.prefix}{n}': ChannelDouble(value=0)
```

Comment:
Why does this have to be ChannelDouble and not pvproperty? I've tried with pvproperty and got an error.

Reply:
Hi @bl0x! It's a good question. The `pvdb` is the lowest-level interface for specifying IOCs in caproto. It maps PV names (strings) to `ChannelData` (and its subclasses, such as `ChannelDouble`).

`PVGroup` and `pvproperty` belong to a higher-level interface that, at the end of the day, also produce a `pvdb` of `ChannelData` objects. But it's generally a nicer interface to use for complex applications.

What you have tried to do is _mix_ the two, and that's not possible. If you are hand-altering the `pvdb` dict, directly you're obliged to use `ChannelData`. If you are using `pvproperty`, you must rely on `PVGroup` to build your `pvdb` for you.